### PR TITLE
NIFI-1118 adding SplitText features of size limit and header marker c…

### PR DESCRIPTION
**Pull request by markobean**
NIFI-1118 - Includes the addition of the new features of maximum split size limitation and header marker character(s) to identify header lines. Also, the recent bug fix to the Remove Trailing New Lines was incorporated.

### CODE REVIEW

**Comment by mosermw** [comment]
Reviewing.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 1]
```diff
@@ -94,12 +118,16 @@
...
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
     public static final PropertyDescriptor REMOVE_TRAILING_NEWLINES = new PropertyDescriptor.Builder()
             .name("Remove Trailing Newlines")
-            .description("Whether to remove newlines at the end of each split file. This should be false if you intend to merge the split files later. If this is set to "
-                + "'true' and a FlowFile is generated that contains only 'empty lines' (i.e., consists only of \r and \n characters), the FlowFile will not be emitted. "
-                + "Note, however, that if the Header Line Count is greater than 0, the resultant FlowFile will never be empty as it will consist of the header lines, so "
-                + "a FlowFile may be emitted that contians only the header lines.")
+            .description("Whether to remove newlines at the end of each split file. This should be false if you intend to merge the split files later.")
```
Comment (mosermw) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: @markobean I think the longer description is still valid and should remain in place.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 2]
```diff
@@ -147,138 +193,200 @@ protected void init(final ProcessorInitializationContext context) {
...
-
         return numLines;
     }
 
-    private EndOfLineMarker locateEndOfLine(final InputStream in, final OutputStream out, final boolean includeLineDelimiter,
-        final EndOfLineBuffer eolBuffer, final byte[] leadingNewLineBytes) throws IOException {
-
+    private EndOfLineMarker countBytesToSplitPoint(final InputStream in, final OutputStream out, final long bytesReadSoFar, final long maxSize,
+                                                   final boolean includeLineDelimiter, final EndOfLineBuffer eolBuffer, final byte[] leadingNewLineBytes) throws IOException {
         int lastByte = -1;
```
Comment (mosermw) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: Looks like the lastByte local var is not used anymore.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 3]
```diff
@@ -147,138 +193,200 @@ protected void init(final ProcessorInitializationContext context) {
...
         long bytesRead = 0L;
+        final ByteArrayOutputStream buffer;
+        if (out != null) {
+            buffer = new ByteArrayOutputStream();
+        } else {
+            buffer = null;
+        }
```
Comment (mosermw) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: I stared at this if-else for longer than I care to admit.  A ternary operator for conditional assignment to a final variable would have helped me, at least, maybe others too.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 4]
```diff
@@ -147,138 +193,200 @@ protected void init(final ProcessorInitializationContext context) {
...
+                return new EndOfLineMarker(-1, eolBuffer, false, leadingNewLineBytes);
             }
 
             // if we have a new line, then we're done
             if (nextByte == '\n') {
-                eolBuffer.addEndOfLine(lastByte == '\r', true);
+                if (buffer != null) {
+                    buffer.writeTo(out);
+                    buffer.close();
+                    eolBuffer.addEndOfLine(false, true);  //TODO: verify "false" is equivalent to "lastByte == '\r'"
```
Comment (mosermw) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: Is this TODO a reminder to future @markobean?  Does a current unit test verify it?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestSplitText.java [hunk 5]
```diff
@@ -39,6 +39,303 @@
...
+        runner.setProperty(SplitText.LINE_SPLIT_COUNT, "2");
+        runner.setProperty(SplitText.FRAGMENT_MAX_SIZE, "20 B");
+
+        runner.enqueue("Line #1\nLine #2\nLine #3\nLong line exceeding limit");
+        runner.run();
+
+        runner.assertTransferCount(SplitText.REL_FAILURE, 0);
+        runner.assertTransferCount(SplitText.REL_ORIGINAL, 1);
+        runner.assertTransferCount(SplitText.REL_SPLITS, 3);
+    }
```
Comment (mosermw) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestSplitText.java]: I examined the content of each output FlowFile.  The first and third look fine, but the second FlowFile had a trailing newline even though the dreaded REMOVE_TRAILING_NEWLINES defaults to "true".

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java (hunk 1 above)
Comment (markobean) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: @mosermw, agreed. Original description put back in place (with spelling correction)

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java (hunk 2 above)
Comment (markobean) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: removed lastByte variable

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java (hunk 4 above)
Comment (markobean) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: Leftover comment with no current significance. Removed.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java (hunk 3 above)
Comment (markobean) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: It made more sense when this method had the possibility of having 'out' be null; this occurred by design in a previous version. Now, it is only called with a valid OutputStream. Still, I think this doesn't hurt and adds a layer of safety should the method be called differently in the future. However, if you still object,  the if/else can be removed and the final buffer be instantiated outright; all unit tests confirmed to pass in this configuration (although not staged to be committed at this time.)

Comment (mosermw) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: No objection to leaving it as-is.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestSplitText.java (hunk 5 above)
Comment (markobean) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestSplitText.java]: Talk about thorough! I don't know how you managed to catch this edge case. However, for completeness, the RTN functionality once again increased complexity and another buffer count was added. Now, if the size limit is exceeded (with RTN = true and no header lines), then the number of EOL bytes previously added to the info.lengthBytes are subtracted from the previous line effectively removing the final EOL characters of the split.

**Comment by mosermw** [comment]
+1 LGTM.  contrib-check on nifi-standard-bundle passes.  Old unit tests from NIFI-1649 continue to work and new unit tests make sense.  Templates from NIFI-1649 continue to work the same.  Will squash, merge and push tomorrow if nobody else reviews.  This is a long standing ticket on a heavily used processor, so I want to give others a chance to weigh in if they want to.

**Pull request closed without merging**